src/Microsoft.DocAsCode.Common/FileAbstractLayer/LinkFileSystem.cs

```diff
@@ -0,0 +1,277 @@
+            {
+                throw new ArgumentNullException(nameof(mappings));
+            }
+            Mappings = mappings.ToImmutableArray();
+            OutputFolder = outputFolder;
+        }
+
+        #endregion
+
+        #region Public Methods
+
+        public ImmutableHashSet<RelativePath> GetAllInputFiles()
+        {
+            EnsureAllInputs();
+            return _allInputs.Keys.ToImmutableHashSet();
+        }
+
+        public ImmutableHashSet<RelativePath> GetAllOutputFiles()
+        {
+            if (IsReadOnly)
+            {
+                throw new InvalidOperationException();
+            }
+            return (from m in _outputList
+                    select m.LogicPath).ToImmutableHashSet();
+        }
+
+        public LinkFileSystem CreateNextLayer(string outputFolder = null) =>
+            new LinkFileSystem(_outputList, outputFolder);
+
+        public bool Exists(string file)
+        {
+            return Exists((RelativePath)file);
+        }
+
+        public bool Exists(RelativePath file)
+        {
+            if (file == null)
+            {
+                throw new ArgumentNullException(nameof(file));
+            }
+            return FindPhysicPathNoThrow(file) != null;
+        }
+
+        public FileStream OpenRead(string file)
+        {
+            return OpenRead((RelativePath)file);
+        }
+
+        public FileStream OpenRead(RelativePath file)
+        {
+            string pp = FindPhysicPath(file.GetPathFromWorkingFolder());
+            return File.OpenRead(pp);
+        }
+
+        public string ReadAllText(string file)
+        {
+            using (var sr = new StreamReader(OpenRead(file)))
+            {
+                return sr.ReadToEnd();
+            }
+        }
+
+        public string ReadAllText(RelativePath file)
+        {
+            using (var sr = new StreamReader(OpenRead(file)))
+            {
+                return sr.ReadToEnd();
+            }
+        }
+
+        public FileStream Create(string file) =>
+            Create((RelativePath)file);
+
+        public FileStream Create(RelativePath file)
+        {
+            if (IsReadOnly)
+            {
+                throw new InvalidOperationException();
+            }
+            var pair = CreateRandomFileStream();
+            _outputList.Add(new PathMapping(file.GetPathFromWorkingFolder(), Path.Combine(OutputFolder, pair.Item1)));
+            return pair.Item2;
+        }
+
+        public void Copy(string sourceFileName, string destFileName) =>
+            Copy((RelativePath)sourceFileName, (RelativePath)destFileName);
+
+        public void Copy(RelativePath sourceFileName, RelativePath destFileName)
+        {
+            string pp = FindPhysicPath(sourceFileName.GetPathFromWorkingFolder());
+            _outputList.Add(new PathMapping(destFileName.GetPathFromWorkingFolder(), pp));
+        }
+
+        #region Help Methods
+
+        public string GetRandomEntry()
```

Comment:
we can remove it from IncrementalUtility when fal is ready.

---
In reply to: [89574561](https://github.com/dotnet/docfx/pull/926#discussion_r89574561) [](ancestors = 89574561)